src/Pos/Client/Txp/History.hs

```diff
@@ -198,65 +206,49 @@ type TxHistoryRedirect =
-        tip <- GS.getTip
-
-        let getGenUtxo = Ether.asks' unGenesisUtxo
-        (bot, genUtxo) <- maybe ((,) <$> GS.getBot <*> getGenUtxo) pure mInit
-
-        -- Getting list of all hashes in main blockchain (excluding bottom block - it's genesis anyway)
-        hashList <- flip unfoldrM tip $ \h ->
-            if h == bot
-            then return Nothing
-            else do
-                header <- DB.runBlockDBRedirect $ DB.blkGetHeader @ssc h >>=
-                    maybeThrow (DBMalformed "Best blockchain is non-continuous")
-                let prev = header ^. prevBlockL
-                return $ Just (h, prev)
-
-        -- Determine last block which txs should be cached
-        let cachedHashes = drop blkSecurityParam hashList
-            nonCachedHashes = take blkSecurityParam hashList
-
+    getBlockHistory
+        :: forall ssc. SscHelpersClass ssc
+        => Tagged ssc ([Address] -> TxHistoryRedirect m (DList TxHistoryEntry))
+    getBlockHistory = Tagged $ \addrs -> do
+        bot <- GS.getBot
         sd <- GS.getSlottingData
 
-        let blockFetcher h txs = do
-                blk <- lift . lift . DB.runBlockDBRedirect $ DB.blkGetBlock @ssc h >>=
-                       maybeThrow (DBMalformed "A block mysteriously disappeared!")
-                deriveAddrHistoryPartialWithTimestamp txs addrs [blk] (Just sd)
-            localFetcher blkTxs = do
-                let mp (txid, TxAux {..}) =
-                      (WithHash taTx txid, taWitness, taDistribution)
-                ltxs <- lift . lift $ getLocalTxs
-                txs <- getRelatedTxs addrs $ map mp ltxs
-                return $ txs ++ blkTxs
-
-        mres <- runMaybeT $ do
-            (cachedTxs, cachedUtxo) <- runUtxoStateT
-                (foldrM blockFetcher [] cachedHashes) genUtxo
-
-            result <- evalUtxoStateT
-                (foldrM blockFetcher cachedTxs nonCachedHashes >>= localFetcher)
-                cachedUtxo
-
-            let lastCachedHash = fromMaybe bot $ head cachedHashes
-            return $ TxHistoryAnswer lastCachedHash (length cachedTxs) cachedUtxo result
-
-        maybe (error "deriveAddrHistory: Nothing") pure mres
+        let fromBlund :: Blund ssc -> GenesisHistoryFetcher m (Block ssc)
+            fromBlund = pure . fst
+
+            getBlockTimestamp :: MainBlock ssc -> Maybe Timestamp
+            getBlockTimestamp blk = getSlotStartPure True (blk ^. mainBlockSlot) sd
+
+            blockFetcher :: HeaderHash -> GenesisHistoryFetcher m (DList TxHistoryEntry)
+            blockFetcher start = GS.foldlUpWhileM fromBlund start (const $ const True)
+                (deriveAddrHistoryBlk addrs getBlockTimestamp) mempty
+
+        DB.runBlockDBRedirect . runGenesisToil . evalToilTEmpty $
+            blockFetcher bot
+
+    getLocalHistory :: [Address] -> TxHistoryRedirect m (DList TxHistoryEntry)
+    getLocalHistory addrs = runDBTxp . evalToilTEmpty $ do
```

Comment:
I don't understand why this function doesn't use `txMempoolToModifier`.
IMHO it's safer and more consistent